homeassistant/components/nest/__init__.py

```diff
@@ -76,76 +80,44 @@
-        _LOGGER.debug("configurator callback")
-        pin = data.get('pin')
-        if await async_setup_nest(hass, nest, config, pin=pin):
-            # start nest update event listener as we missed startup hook
-            hass.async_add_job(async_nest_update_event_broker, hass, nest)
-
-    _CONFIGURING['nest'] = configurator.async_request_config(
-        "Nest", async_nest_config_callback,
-        description=('To configure Nest, click Request Authorization below, '
-                     'log into your Nest account, '
-                     'and then enter the resulting PIN'),
-        link_name='Request Authorization',
-        link_url=nest.authorize_url,
-        submit_caption="Confirm",
-        fields=[{'id': 'pin', 'name': 'Enter the PIN', 'type': ''}]
-    )
-
-
-async def async_setup_nest(hass, nest, config, pin=None):
-    """Set up the Nest devices."""
-    from nest.nest import AuthorizationError, APIError
-    if pin is not None:
-        _LOGGER.debug("pin acquired, requesting access token")
-        error_message = None
-        try:
-            nest.request_token(pin)
-        except AuthorizationError as auth_error:
-            error_message = "Nest authorization failed: {}".format(auth_error)
-        except APIError as api_error:
-            error_message = "Failed to call Nest API: {}".format(api_error)
-
-        if error_message is not None:
-            _LOGGER.warning(error_message)
-            hass.components.configurator.async_notify_errors(
-                _CONFIGURING['nest'], error_message)
-            return False
-
-    if nest.access_token is None:
-        _LOGGER.debug("no access_token, requesting configuration")
-        await async_request_configuration(nest, hass, config)
-        return False
+    conf = config[DOMAIN]
+
+    local_auth.initialize(hass, conf[CONF_CLIENT_ID], conf[CONF_CLIENT_SECRET])
+
+    filename = config.get(CONF_FILENAME, NEST_CONFIG_FILE)
+    access_token_cache_file = hass.config.path(filename)
+
+    if await hass.async_add_job(os.path.isfile, access_token_cache_file):
+        hass.async_add_job(hass.config_entries.flow.async_init(
+            DOMAIN, source='import', data={
+                'nest_conf_path': access_token_cache_file,
+            }
+        ))
+
+    # Store config to be used during entry setup
+    hass.data[DATA_NEST_CONFIG] = conf
+
+    return True
+
+
+async def async_setup_entry(hass, entry):
+    """Setup Nest from a config entry."""
+    from nest import Nest
 
-    if 'nest' in _CONFIGURING:
-        _LOGGER.debug("configuration done")
-        configurator = hass.components.configurator
-        configurator.async_request_done(_CONFIGURING.pop('nest'))
+    nest = Nest(access_token=entry.data['tokens']['access_token'])
 
     _LOGGER.debug("proceeding with setup")
-    conf = config[DOMAIN]
+    conf = hass.data.get(DATA_NEST_CONFIG, {})
     hass.data[DATA_NEST] = NestDevice(hass, conf, nest)
```

Comment:
There may have a potential issue, inside NestDevice constructor, it will first call nest.Nest._status, which will trigger the establish of the persistence HTTP connection. This function may block event loop for few seconds, depends on network condition. It wouldn't be necessary address in this PR. I will figure out some solution, or take this chane to change python-nest better support asyncio.

Reply:
Good catch! We should never do I/O inside the event loop. I've pushed the I/O pieces into a standalone initialize function.